design-proposals/sidecar-hooks.md

```diff
@@ -0,0 +1,54 @@
+# Sidecar hooks
+
+## Overview
+
+A hooking mechanism that allows customizations to the VMI before it starts without changing KubeVirt
+core components. The common user of those hooks are Sidecars that will run in the same Pod that the
+VMI will be deployed.
+
+## Motivation
+
+A way to accomodate use cases that we might not want to bring and support in KubeVirt itself or to
+give room to test and develop features while they are still not mature enough.
+
+## Goals
+
+- Provide and maintain a set of gRPC APIs for users to do modifications to the VMI
+
+## Non Goals
+
+- Maintain a wide variety of user's applications in KubeVirt codebase
```

Comment:
Maybe expand on the non-goals: What types of user applications are considered "wide variety," and why is it not a goal to maintain them in the KubeVirt codebase?

Reply:
That's a good point. I think the important factor is core functionalities for KubeVirt that can be helped with sidecars. We do provide and maintain Today sidecars that configure slirp and passt, named "network pluggins".

I think it is reasonable that we improve hooks API as needed for core functionalities like that, but not necessarily to all cases.

If we do have a lot of sidecars with custom features that users want to have in a single place, we should simply create another repository for it. It does not need to be in KubeVirt repo, nor maintained by KubeVirt developers.